File: samples/csharp/common/AutoML/ConsoleHelper.cs
using System;
using System.Linq;
using System.Collections.Generic;
using Microsoft.ML.Data;
using Microsoft.ML;
using Microsoft.ML.AutoML;
using System.Text;

namespace Common
{
    public static class ConsoleHelper
    {
        private const int Width = 114;

        public static void PrintRegressionMetrics(string name, RegressionMetrics metrics)
        {
            Console.WriteLine($"*************************************************");
            Console.WriteLine($"*       Metrics for {name} regression model      ");
            Console.WriteLine($"*------------------------------------------------");
            Console.WriteLine($"*       LossFn:        {metrics.LossFunction:0.##}");
            Console.WriteLine($"*       R2 Score:      {metrics.RSquared:0.##}");
            Console.WriteLine($"*       Absolute loss: {metrics.MeanAbsoluteError:#.##}");
            Console.WriteLine($"*       Squared loss:  {metrics.MeanSquaredError:#.##}");
            Console.WriteLine($"*       RMS loss:      {metrics.RootMeanSquaredError:#.##}");
            Console.WriteLine($"*************************************************");
        }

        public static void PrintBinaryClassificationMetrics(string name, BinaryClassificationMetrics metrics)
        {
            Console.WriteLine($"************************************************************");
            Console.WriteLine($"*       Metrics for {name} binary classification model      ");
            Console.WriteLine($"*-----------------------------------------------------------");
            Console.WriteLine($"*       Accuracy: {metrics.Accuracy:P2}");
            Console.WriteLine($"*       Area Under Curve:      {metrics.AreaUnderRocCurve:P2}");
            Console.WriteLine($"*       Area under Precision recall Curve:  {metrics.AreaUnderPrecisionRecallCurve:P2}");
            Console.WriteLine($"*       F1Score:  {metrics.F1Score:P2}");
            Console.WriteLine($"*       PositivePrecision:  {metrics.PositivePrecision:#.##}");
            Console.WriteLine($"*       PositiveRecall:  {metrics.PositiveRecall:#.##}");
            Console.WriteLine($"*       NegativePrecision:  {metrics.NegativePrecision:#.##}");
            Console.WriteLine($"*       NegativeRecall:  {metrics.NegativeRecall:P2}");
            Console.WriteLine($"************************************************************");
        }

        public static void PrintMulticlassClassificationMetrics(string name, MulticlassClassificationMetrics metrics)
        {
            Console.WriteLine($"************************************************************");
            Console.WriteLine($"*    Metrics for {name} multi-class classification model   ");
            Console.WriteLine($"*-----------------------------------------------------------");
            Console.WriteLine($"    MacroAccuracy = {metrics.MacroAccuracy:0.####}, a value between 0 and 1, the closer to 1, the better");
            Console.WriteLine($"    MicroAccuracy = {metrics.MicroAccuracy:0.####}, a value between 0 and 1, the closer to 1, the better");
            Console.WriteLine($"    LogLoss = {metrics.LogLoss:0.####}, the closer to 0, the better");
            Console.WriteLine($"    LogLoss for class 1 = {metrics.PerClassLogLoss[0]:0.####}, the closer to 0, the better");
            Console.WriteLine($"    LogLoss for class 2 = {metrics.PerClassLogLoss[1]:0.####}, the closer to 0, the better");
            Console.WriteLine($"    LogLoss for class 3 = {metrics.PerClassLogLoss[2]:0.####}, the closer to 0, the better");
            Console.WriteLine($"************************************************************");
        }

        public static void ShowDataViewInConsole(MLContext mlContext, IDataView dataView, int numberOfRows = 4)
        {
            string msg = string.Format("Show data in DataView: Showing {0} rows with the columns", numberOfRows.ToString());
            ConsoleWriteHeader(msg);

            var preViewTransformedData = dataView.Preview(maxRows: numberOfRows);

            foreach (var row in preViewTransformedData.RowView)
            {
                var ColumnCollection = row.Values;
                string lineToPrint = "Row--> ";
                foreach (KeyValuePair<string, object> column in ColumnCollection)
                {
                    lineToPrint += $"| {column.Key}:{column.Value}";
                }
                Console.WriteLine(lineToPrint + "\n");
            }
        }

        internal static void PrintIterationMetrics(int iteration, string trainerName, BinaryClassificationMetrics metrics, double? runtimeInSeconds)
        {
            CreateRow($"{iteration,-4} {trainerName,-35} {metrics?.Accuracy ?? double.NaN,9:F4} {metrics?.AreaUnderRocCurve ?? double.NaN,8:F4} {metrics?.AreaUnderPrecisionRecallCurve ?? double.NaN,8:F4} {metrics?.F1Score ?? double.NaN,9:F4} {runtimeInSeconds.Value,9:F1}", Width);
        }

        internal static void PrintIterationMetrics(int iteration, string trainerName, MulticlassClassificationMetrics metrics, double? runtimeInSeconds)
        {
            CreateRow($"{iteration,-4} {trainerName,-35} {metrics?.MicroAccuracy ?? double.NaN,14:F4} {metrics?.MacroAccuracy ?? double.NaN,14:F4} {runtimeInSeconds.Value,9:F1}", Width);
        }

        internal static void PrintIterationMetrics(int iteration, string trainerName, RegressionMetrics metrics, double? runtimeInSeconds)
        {
            CreateRow($"{iteration,-4} {trainerName,-35} {metrics?.RSquared ?? double.NaN,8:F4} {metrics?.MeanAbsoluteError ?? double.NaN,13:F2} {metrics?.MeanSquaredError ?? double.NaN,12:F2} {metrics?.RootMeanSquaredError ?? double.NaN,8:F2} {runtimeInSeconds.Value,9:F1}", Width);
        }

        internal static void PrintIterationException(Exception ex)
        {
            Console.WriteLine($"Exception during AutoML iteration: {ex}");
        }

        internal static void PrintBinaryClassificationMetricsHeader()
        {
            CreateRow($"{"",-4} {"Trainer",-35} {"Accuracy",9} {"AUC",8} {"AUPRC",8} {"F1-score",9} {"Duration",9}", Width);
        }

        internal static void PrintMulticlassClassificationMetricsHeader()
        {
            CreateRow($"{"",-4} {"Trainer",-35} {"MicroAccuracy",14} {"MacroAccuracy",14} {"Duration",9}", Width);
        }

        internal static void PrintRegressionMetricsHeader()
        {
            CreateRow($"{"",-4} {"Trainer",-35} {"RSquared",8} {"Absolute-loss",13} {"Squared-loss",12} {"RMS-loss",8} {"Duration",9}", Width);
        }

        private static void CreateRow(string message, int width)
        {
            Console.WriteLine("|" + message.PadRight(width - 2) + "|");
        }

        public static void ConsoleWriteHeader(params string[] lines)
        {
            var defaultColor = Console.ForegroundColor;
            Console.ForegroundColor = ConsoleColor.Yellow;
            Console.WriteLine(" ");
            foreach (var line in lines)
            {
                Console.WriteLine(line);
            }
            var maxLength = lines.Select(x => x.Length).Max();
            Console.WriteLine(new string('#', maxLength));
            Console.ForegroundColor = defaultColor;
        }

        public static void Print(ColumnInferenceResults results)
        {
            Console.WriteLine("Inferred dataset columns --");
            new ColumnInferencePrinter(results).Print();
            Console.WriteLine();
        }

        public static string BuildStringTable(IList<string[]> arrValues)
        {
            int[] maxColumnsWidth = GetMaxColumnsWidth(arrValues);
            var headerSpliter = new string('-', maxColumnsWidth.Sum(i => i + 3) - 1);

            var sb = new StringBuilder();
            for (int rowIndex = 0; rowIndex < arrValues.Count; rowIndex++)
            {
                if (rowIndex == 0)
                {
                    sb.AppendFormat("  {0} ", headerSpliter);
                    sb.AppendLine();
                }

                for (int colIndex = 0; colIndex < arrValues[0].Length; colIndex++)
                {
                    // Print cell
                    string cell = arrValues[rowIndex][colIndex];
                    cell = cell.PadRight(maxColumnsWidth[colIndex]);
                    sb.Append(" | ");
                    sb.Append(cell);
                }

                // Print end of line
                sb.Append(" | ");
                sb.AppendLine();

                // Print splitter
                if (rowIndex == 0)
                {
                    sb.AppendFormat(" |{0}| ", headerSpliter);
                    sb.AppendLine();
                }

                if (rowIndex == arrValues.Count - 1)
                {
                    sb.AppendFormat("  {0} ", headerSpliter);
                }
            }

            return sb.ToString();
        }

        private static int[] GetMaxColumnsWidth(IList<string[]> arrValues)
        {
            var maxColumnsWidth = new int[arrValues[0].Length];
            for (int colIndex = 0; colIndex < arrValues[0].Length; colIndex++)
            {
                for (int rowIndex = 0; rowIndex < arrValues.Count; rowIndex++)
                {
                    int newLength = arrValues[rowIndex][colIndex].Length;
                    int oldLength = maxColumnsWidth[colIndex];

                    if (newLength > oldLength)
                    {
                        maxColumnsWidth[colIndex] = newLength;
                    }
                }
            }

            return maxColumnsWidth;
        }

        class ColumnInferencePrinter
        {
            private static readonly string[] TableHeaders = new[] { "Name", "Data Type", "Purpose" };

            private readonly ColumnInferenceResults _results;

            public ColumnInferencePrinter(ColumnInferenceResults results)
            {
                _results = results;
            }

            public void Print()
            {
                var tableRows = new List<string[]>();

                // Add headers
                tableRows.Add(TableHeaders);

                // Add column data
                var info = _results.ColumnInformation;
                AppendTableRow(tableRows, info.LabelColumnName, "Label");
                AppendTableRow(tableRows, info.ExampleWeightColumnName, "Weight");
                AppendTableRow(tableRows, info.SamplingKeyColumnName, "Sampling Key");
                AppendTableRows(tableRows, info.CategoricalColumnNames, "Categorical");
                AppendTableRows(tableRows, info.NumericColumnNames, "Numeric");
                AppendTableRows(tableRows, info.TextColumnNames, "Text");
                AppendTableRows(tableRows, info.IgnoredColumnNames, "Ignored");

                Console.WriteLine(ConsoleHelper.BuildStringTable(tableRows));
            }

            private void AppendTableRow(ICollection<string[]> tableRows,
                string columnName, string columnPurpose)
            {
                if (columnName == null)
                {
                    return;
                }

                tableRows.Add(new[]
                {
                columnName,
                GetColumnDataType(columnName),
                columnPurpose
            });
            }

            private void AppendTableRows(ICollection<string[]> tableRows,
                IEnumerable<string> columnNames, string columnPurpose)
            {
                foreach (var columnName in columnNames)
                {
                    AppendTableRow(tableRows, columnName, columnPurpose);
                }
            }

            private string GetColumnDataType(string columnName)
            {
                return _results.TextLoaderOptions.Columns.First(c => c.Name == columnName).DataKind.ToString();
            }
        }
    }
}

File: samples/fsharp/common/AutoML/ConsoleHelper.fs
module Common.ConsoleHelper

open System
open Microsoft.ML
open Microsoft.ML.Data
open Microsoft.ML.AutoML

let width = 114

let printRegressionMetrics name (metrics : RegressionMetrics) =
    printfn "*************************************************"
    printfn "*       Metrics for {name} regression model      "
    printfn "*------------------------------------------------"
    printfn "*       LossFn:        %.2f" metrics.LossFunction
    printfn "*       R2 Score:      %.2f" metrics.RSquared
    printfn "*       Absolute loss: %.2f" metrics.MeanAbsoluteError
    printfn "*       Squared loss:  %.2f" metrics.MeanSquaredError
    printfn "*       RMS loss:      %.2f" metrics.RootMeanSquaredError
    printfn "*************************************************"

let printBinaryClassificationMetrics name (metrics : BinaryClassificationMetrics) =
    printfn"************************************************************"
    printfn"*       Metrics for %s binary classification model      " name
    printfn"*-----------------------------------------------------------"
    printfn"*       Accuracy: %.2f%%" (metrics.Accuracy * 100.)
    printfn"*       Area Under Curve:      %.2f%%" (metrics.AreaUnderRocCurve * 100.)
    printfn"*       Area under Precision recall Curve:    %.2f%%" (metrics.AreaUnderPrecisionRecallCurve * 100.)
    printfn"*       F1Score:  %.2f%%" (metrics.F1Score * 100.)
    printfn"*       PositivePrecision:      %.2f" (metrics.PositivePrecision)
    printfn"*       PositiveRecall:      %.2f" (metrics.PositiveRecall)
    printfn"*       NegativePrecision:      %.2f" (metrics.NegativePrecision)
    printfn"*       NegativeRecall:      %.2f" (metrics.NegativeRecall)
    printfn"************************************************************"

let printMultiClassClassificationMetrics name (metrics : MulticlassClassificationMetrics) =
    printfn "************************************************************"
    printfn "*    Metrics for %s multi-class classification model   " name
    printfn "*-----------------------------------------------------------"
    printfn "    AccuracyMacro = %.4f, a value between 0 and 1, the closer to 1, the better" metrics.MacroAccuracy
    printfn "    AccuracyMicro = %.4f, a value between 0 and 1, the closer to 1, the better" metrics.MacroAccuracy
    printfn "    LogLoss = %.4f, the closer to 0, the better" metrics.LogLoss
    printfn "    LogLoss for class 1 = %.4f, the closer to 0, the better" metrics.PerClassLogLoss.[0]
    printfn "    LogLoss for class 2 = %.4f, the closer to 0, the better" metrics.PerClassLogLoss.[1]
    printfn "    LogLoss for class 3 = %.4f, the closer to 0, the better" metrics.PerClassLogLoss.[2]
    printfn "************************************************************"


let createRow width (message : string) = sprintf "|%s|" (message.PadRight(width - 2))

let printRow l = 
    l
    |> List.map (fun (p,msg : string) -> if p > 0 then msg.PadLeft(p, ' ') else msg.PadRight(-p, ' '))
    |> String.concat " "
    |> createRow width 
    |> printfn "%s"
    

let regressionMetricsHeader = 
    [
        -4, ""
        -35, "Trainer"
        8, "RSquared"
        13, "Absolute-loss"
        12, "Squared-loss"
        8, "RMS-loss"
        9, "Duration"
    ]

let printRegressionMetricsHeader() = printRow regressionMetricsHeader

let printRegressionIterationMetrics iteration trainerName (metrics : RegressionMetrics) runtimeInSeconds = 
    [
        string iteration
        trainerName
        sprintf "%0.4f" metrics.RSquared
        sprintf "%0.2f" metrics.MeanAbsoluteError
        sprintf "%0.2f" metrics.MeanSquaredError
        sprintf "%0.2f" metrics.RootMeanSquaredError
        sprintf "%0.1f" runtimeInSeconds
    ]
    |> List.zip (regressionMetricsHeader |> List.map fst)
    |> printRow

let binaryMetricsHeader = 
    [
        -4, ""
        -35, "Trainer"
        9, "Accuracy"
        8, "AUC"
        8, "AUPRC"
        9, "F1-score"
        9, "Duration"
    ]

let printBinaryMetricsHeader() = printRow binaryMetricsHeader

let printBinaryIterationMetrics iteration trainerName (metrics : BinaryClassificationMetrics) runtimeInSeconds = 
    [
        string iteration
        trainerName
        sprintf "%0.4f" metrics.Accuracy
        sprintf "%0.2f" metrics.AreaUnderRocCurve
        sprintf "%0.2f" metrics.AreaUnderPrecisionRecallCurve
        sprintf "%0.2f" metrics.F1Score
        sprintf "%0.1f" runtimeInSeconds
    ]
    |> List.zip (binaryMetricsHeader |> List.map fst)
    |> printRow

let multiclassMetricsHeader = 
    [
        -4, ""
        -35, "Trainer"
        14, "MicroAccuracy"
        14, "MacroAccuracy"
        9, "Duration"
    ]

let printMulticlassMetricsHeader() = printRow multiclassMetricsHeader

let printMulticlassIterationMetrics iteration trainerName (metrics : MulticlassClassificationMetrics) runtimeInSeconds = 
    [
        string iteration
        trainerName
        sprintf "%0.4f" metrics.MicroAccuracy
        sprintf "%0.2f" metrics.MacroAccuracy
        sprintf "%0.1f" runtimeInSeconds
    ]
    |> List.zip (multiclassMetricsHeader |> List.map fst)
    |> printRow

let print (results : ColumnInferenceResults) = 
    let dataTypes = results.TextLoaderOptions.Columns |> Seq.map (fun x -> x.Name, string x.DataKind) |> dict
    let row purpose name = 
        match name with 
        | null -> None 
        | _ -> Some [name; dataTypes.[name]; purpose]
    let header = ["Name"; "Data Type"; "Purpose"]
    let rows = 
        [
            yield Some header
            yield row "Label" results.ColumnInformation.LabelColumnName 
            yield row "Weight" results.ColumnInformation.ExampleWeightColumnName 
            yield row "Sampling Key" results.ColumnInformation.SamplingKeyColumnName
            yield! results.ColumnInformation.CategoricalColumnNames |> Seq.map (row "Categorical")
            yield! results.ColumnInformation.NumericColumnNames |> Seq.map (row "Numeric")
            yield! results.ColumnInformation.TextColumnNames |> Seq.map (row "Text")
            yield! results.ColumnInformation.IgnoredColumnNames |> Seq.map (row "Ignored")
        ]
        |> List.choose id
    let lengths = List.init 3 (fun i -> rows |> List.map (fun x -> x.[i].Length) |> List.max)
    let rowLength = 
        let length = lengths |> Seq.sum
        length + 8
    let bar = String.replicate rowLength "-" |> sprintf "  %s"
    let innerBar = String.replicate rowLength "-" |> sprintf " |%s|"
    let fmtRow (l : string list) = 
        let l = l |> List.mapi (fun i x -> x.PadRight(lengths.[i], ' '))
        sprintf " | %s | %s | %s |" l.[0] l.[1] l.[2]
    [
        yield "Inferred dataset columns --"
        yield bar
        yield fmtRow header
        yield innerBar
        yield! rows.Tail |> List.map fmtRow
        yield bar
        yield ""
    ]
    |> List.iter (printfn "%s")

let consoleWriteHeader line =
    let defaultColor = Console.ForegroundColor
    Console.ForegroundColor <- ConsoleColor.Yellow
    printfn " "
    printfn "%s" line
    let maxLength = line.Length
    printfn "%s" (new string('#', maxLength))
    Console.ForegroundColor <- defaultColor

let downcastPipeline (pipeline : IEstimator<'a>) =
    match pipeline with
    | :? IEstimator<ITransformer> as p -> p
    | _ -> failwith "The pipeline has to be an instance of IEstimator<ITransformer>."

let showDataViewInConsole (mlContext : MLContext) (dataView : IDataView) numberOfRows =
    
    let msg = sprintf "Show data in DataView: Showing %d rows with the columns" numberOfRows
    consoleWriteHeader msg

    dataView.Preview(numberOfRows).RowView
    |> Seq.iter 
        (fun row ->
            row.Values
            |> Array.map (function KeyValue(k,v) -> sprintf "| %s:%O" k v)
            |> Array.fold (+) "Row--> "
            |> printfn "%s\n"
        )


let printIterationException (ex : exn) = 
    printf "Exception during AutoML iteration: %O" ex

let progressHandler printHeader printIterMetrics =
    let mutable iterIndex = 0
    {new IProgress<RunDetail<'a>> with
         member this.Report(value: RunDetail<'a>): unit = 
            if iterIndex = 0 then 
                printHeader()
            iterIndex <- iterIndex + 1
            match value.Exception with 
            | null -> 
                printIterMetrics iterIndex value.TrainerName value.ValidationMetrics value.RuntimeInSeconds
            | ex -> 
                printIterationException ex
    }

let regressionExperimentProgressHandler() = progressHandler printRegressionMetricsHeader printRegressionIterationMetrics
let binaryExperimentProgressHandler() = progressHandler printBinaryMetricsHeader printBinaryIterationMetrics
let multiclassExperimentProgressHandler() = progressHandler printMulticlassMetricsHeader printMulticlassIterationMetrics
